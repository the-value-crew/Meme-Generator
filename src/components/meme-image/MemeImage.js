import React from 'react';
import { fabric } from "fabric";

let fabricObject = {};
let canvas;

export default class MemeImage extends React.Component {

    componentDidMount() {
        canvas = new fabric.Canvas('memeCanvas');
    }

    state = {
        memeTextObject: {},
    }

    static getDerivedStateFromProps(props, state) {

        let stateChanges = {}

        if (props.memeTextObject !== state.memeTextObject) {
            stateChanges.memeTextObject = props.memeTextObject
        }

        return stateChanges;
    }

    componentDidUpdate(prevPrps, prevState) {

        console.log(this.state.memeTextObject);
        let memeTextObject = this.state.memeTextObject;

        this.loadCanvasData(canvas);
        
        if (this.props.memeBackgroundImage) {
            this.loadBackgroundImageToCanvas(canvas, this.props.memeBackgroundImage.url);
        }

        if (memeTextObject.text !== "" && prevPrps.memeTextObject.text !== this.state.memeTextObject.text) {
            this.addTextToCanvas(canvas, memeTextObject);
        }

        this.saveCanvasData(canvas);

        if (this.props.downloadCanvas) {
            this.setState({ memeTextObject: {} });
            this.props.onMemeTextClear();
            this.saveImage(canvas);
            this.deleteCanvasObjects(canvas);
        }

        if (this.props.resetCanvas) {
            this.deleteCanvasObjects(canvas);
            this.props.onCanvasReset();
        }

        canvas.on('mouse:out', (e) => this.handleMouseOut(e, canvas));

    }


    saveCanvasData(canvas) {
        fabricObject = JSON.stringify(canvas.toJSON());
    }

    deleteCanvasObjects(canvas) {
        fabricObject = JSON.stringify(canvas.remove(...canvas.getObjects()));
    }

    saveImage = (canvas) => {
        let link = document.createElement("a");
        link.href = canvas.toDataURL({
            format: 'jpeg',
            quality: 0.8
        });
        link.download = 'canvas.png';
        link.click();
        this.onImageSaved();
    }

    onImageSaved = () => {
        this.props.onImageDownloaded(true);
    }


    handleMouseOut(e, canvas) {
        fabricObject = this.saveCanvasData(canvas);
        canvas.renderAll();
    }

    loadCanvasData = (canvas) => {
        canvas.loadFromJSON(fabricObject, function () {
            canvas.renderAll();
        });
    }

    addTextToCanvas(canvas, memeTextObject) {
        console.log("Chaliraxa add text");
        canvas.add(new fabric.Text(memeTextObject.text, {
            fontFamily: memeTextObject.settings.fontFamily,
            fill: memeTextObject.settings.color,
            left: 100,
            top: 100
        }));
    }


    loadBackgroundImageToCanvas(canvas, url) {

        fabric.Image.fromURL(url, function (meme) {
            canvas.setWidth(400);
            canvas.setHeight(meme.height);
            canvas.setBackgroundImage(meme, canvas.renderAll.bind(canvas), {
                scaleX: canvas.width / meme.width,
                scaleY: canvas.height / meme.height
            });
        }.bind(this), {
            crossOrigin: 'anonymous'
        });
    }


    render() {
        return (
            <div>
                <canvas id="memeCanvas" style={{ width: "100%", height: "auto" }} />
            </div>
        )
    }
}
