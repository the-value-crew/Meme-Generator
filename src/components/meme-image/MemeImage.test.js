import { shallow, render } from 'enzyme';
import React from 'react';
import MemeImage from './MemeImage';

let shallowWrapper;
let renderWrapper;
beforeEach(() => {
    const mockProps = {
        meme: {
            "id": "129242436",
            "name": "Change My Mind",
            "url": "https://i.imgflip.com/24y43o.jpg",
            "width": 482,
            "height": 361,
            "box_count": 2
        },
        memeText: "Come to the dark side"
    }
    shallowWrapper = shallow(<MemeImage {...mockProps} />);
    renderWrapper = render(<MemeImage {...mockProps} />);
})

it('expects to render meme image component', () => {
    expect(renderWrapper).toMatchSnapshot();
});
