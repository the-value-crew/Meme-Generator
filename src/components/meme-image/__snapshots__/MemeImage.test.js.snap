// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expects to render meme image component 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "data-meme": "https://i.imgflip.com/24y43o.jpg",
          "id": "memeCanvas",
          "style": "width:100%;height:auto",
        },
        "children": Array [],
        "name": "canvas",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "data-meme": undefined,
          "id": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "data-meme": undefined,
          "id": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
